lib/tools/system-calls.js

```diff
@@ -216,4 +217,178 @@ systemCallMethods.getPortFromEmulatorString = function (emStr) {
+    let emulators = await this.getConnectedEmulators();
+    if (emulators.length < 1) {
+      log.errorAndThrow("No emulators connected");
+    }
+    for (let emulator of emulators) {
+      this.setEmulatorPort(emulator.port);
+      let runningAVDName = await this.sendTelnetCommand("avd name");
+      if (avdName === runningAVDName) {
+        log.debug(`Found emulator ${avdName} in port ${emulator.port}`);
+        this.setDeviceId(emulator.udid);
+        return emulator;
+      }
+    }
+    log.debug(`Emulator ${avdName} not running`);
+    return null;
+  } catch (e) {
+    log.errorAndThrow(`Error getting AVD. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getRunningAVDWithRetry = async function (avdName, timeoutMs = 20000) {
+  try {
+    let start = Date.now();
+    while ((Date.now() - start) < timeoutMs) {
+      try {
+        let runningAVD = await this.getRunningAVD(avdName.replace('@', ''));
+        if (runningAVD) {
+          return runningAVD;
+        }
+      } catch (e) {
+        // Do nothing.
+        log.info('Trying to get running AVD');
+      }
+      // cool down
+      await sleep(200);
+    }
+    log.errorAndThrow(`Could not find ${avdName} emulator.`);
+  } catch (e) {
+    log.errorAndThrow(`Error getting AVD with retry. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.killAllEmulators = async function () {
+  let cmd, args;
+  if (system.isWindows()) {
+    cmd = 'TASKKILL';
+    args = ['TASKKILL' ,'/IM', 'emulator.exe'];
+  } else {
+    cmd = '/usr/bin/killall';
+    args = ['-m', 'emulator*'];
+  }
+  try {
+    await exec(cmd, args);
+  } catch (e) {
+    log.errorAndThrow(`Error killing emulators. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.launchAVD = async function (avdName, avdArgs, language, locale,
+  avdLaunchTimeout = 60000, avdReadyTimeout = 60000, retryTimes = 1) {
+  log.debug(`Launching Emulator with AVD ${avdName}, launchTimeout` +
+            `${avdLaunchTimeout} ms and readyTimeout ${avdReadyTimeout} ms`);
+  let emulatorBinaryPath = await this.getSdkBinaryPath("emulator");
+  if (avdName[0] === "@") {
+    avdName = avdName.substr(1);
+  }
+  let launchArgs = ["-avd", avdName];
+  if (typeof language === "string") {
+    log.debug(`Setting Android Device Language to ${language}`);
+    launchArgs.push("-prop", `persist.sys.language=${language.toLowerCase()}`);
+  }
+  if (typeof locale === "string") {
+    log.debug(`Setting Android Device Country to ${locale}`);
+    launchArgs.push("-prop", `persist.sys.country=${locale.toUpperCase()}`);
+  }
+  if (typeof avdArgs === "string") {
+    avdArgs = avdArgs.split(" ");
+    launchArgs = launchArgs.concat(avdArgs);
+  }
+  let proc = child_process.spawn(emulatorBinaryPath, launchArgs);
+  proc.on("error", function (err) {
+    log.error("Unable to start Emulator: " + err.message);
+    // actual error will get caught by getRunningAVDWithRetry
```

Comment:
how would this error get caught by getRunningAVDWithRetry? it's not thrown anywhere...